libs/ui/src/hooks/use_smartcard_auth.ts

```diff
@@ -78,45 +128,68 @@ function cardDataToUser(cardData: AnyCardData): User {
   }
 }
 
-function parseUserFromCard(card: CardApiReady): User | undefined {
-  if (!card.shortValue) return undefined;
-  const cardData = safeParseJson(card.shortValue, AnyCardDataSchema).ok();
-  return cardData && cardDataToUser(cardData);
-}
-
 function attemptLogin(
+  previousAuth: AuthState,
   card: CardApiReady,
   allowedUserRoles: UserRole[],
-  electionDefinition?: ElectionDefinition
+  scope: AuthScope
 ): Result<User, LoggedOutAuth['reason']> {
   const user = parseUserFromCard(card);
   if (!user) return err('invalid_user_on_card');
   if (!allowedUserRoles.includes(user.role)) {
     return err('user_role_not_allowed');
   }
+
   if (user.role === 'pollworker') {
-    if (!electionDefinition) return err('machine_not_configured');
-    if (user.electionHash !== electionDefinition.electionHash) {
-      return err('pollworker_election_hash_mismatch');
+    if (!scope.electionDefinition) return err('machine_not_configured');
+    if (user.electionHash !== scope.electionDefinition.electionHash) {
+      return err('pollworker_wrong_election');
+    }
+  }
+
+  if (user.role === 'voter') {
+    if (!scope.electionDefinition || !scope.precinct) {
+      return err('machine_not_configured');
+    }
+    if (utcTimestamp() >= user.createdAt + VOTER_CARD_EXPIRATION_SECONDS) {
+      return err('voter_card_expired');
+    }
+    if (user.voidedAt) {
+      return err('voter_card_voided');
+    }
+    // After a voter's ballot is printed, we don't want to log them out until
+    // they remove their card
+    if (user.ballotPrintedAt && previousAuth.status !== 'logged_in') {
+      return err('voter_card_printed');
+    }
+    const ballotStyle = getBallotStyle({
+      election: scope.electionDefinition.election,
+      ballotStyleId: user.ballotStyleId,
+    });
+    const precinct = getPrecinctById({
+      election: scope.electionDefinition.election,
+      precinctId: user.precinctId,
+    });
+    if (!ballotStyle || !precinct) {
+      return err('voter_wrong_election');
+    }
```

Comment:
Fuzzy question: Why do we not also store/check election hash for voter cards, as we do for poll worker cards?

Reply:
I'm not sure! Maybe a question for @benadida or @eventualbuddha

Reply:
historical artifact: we haven't gone back to voter cards in a while. We should absolutely include election hash on those at some point.

Reply:
https://github.com/votingworks/vxsuite/issues/2000